_posts/2024-05-27-A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12.md

```diff
@@ -0,0 +1,144 @@
+
+For the bi-encoder mode, the latency unrelated to the inverted index ($$b$$) is 265.4 ms and the time complexity of searching the inverted index ($$k$$) is 20.58 ms/1M docs. The fixed term ($$b$$) is much larger than that of BM25 because to model inference. You can speed up the model inference by using a GPU endpoint, as we show in [Accelerating model inference using GPU](#accelerating-model-inference-using-gpu). Additionally, because of query expansion, the **retrieval latency for the bi-encoder mode is about 4x that of the doc-only mode**.
+
+Note that the exact slope value depends on the distribution of the data. That means for different datasets, we may get different slopes. However, the relative difference between these slopes is likely to remain small. This is because the key factor influencing the slope is token expansion rate, which reflects the model's ability to identify semantically related terms. This ability, ideally, should be consistent across datasets. The experiment results on MS MARCO serve as a strong reference point and can be used to estimate the latency of neural sparse search for other datasets.
+
+For medium to large datasets (millions to tens of millions of documents), neural sparse search is very efficient, with smaller client-side latency and significantly lower resource consumption compared to dense vector search (see [this blog](https://opensearch.org/blog/improving-document-retrieval-with-sparse-semantic-encoders/)). The doc-only mode delivers a better client-side search latency, faster than the time required for machine learning (ML) inference. Neural sparse search provides additional cost savings because you don't need to host a powerful ML service that supports a high throughput. 
+
+However, for extremely large datasets (billions of documents), the latency of neural sparse search also grows linearly with the number of documents. In this case, a dense vector search is a better choice for low latency, because the latency for the underlying Approximate Nearest Neighbor (ANN) algorithm used in dense vector search grows logarithmically with the number of documents. We're actively working on improvements to neural sparse search to address its scalability limitations for extremely large datasets.
```

Comment:
```suggestion
However, for extremely large datasets (billions of documents), the latency of neural sparse search also grows linearly with the number of documents. In this case, a dense vector search is a better choice for low latency because the latency of the underlying approximate nearest neighbor (ANN) algorithm used in dense vector search grows logarithmically with the number of documents. We're actively working on improvements to neural sparse search to address its scalability limitations with extremely large datasets.
```